ecc/bls12-377/fr/gkr/registry.go

```diff
@@ -0,0 +1,301 @@
+// RegisterGate will return an error if it cannot verify that this claim is correct.
+func WithSolvableVar(solvableVar int) RegisterGateOption {
+	return func(settings *registerGateSettings) {
+		settings.solvableVar = solvableVar
+	}
+}
+
+// WithUnverifiedSolvableVar sets the index of a variable whose value can be uniquely determined from that of the other variables along with the gate's output.
+// RegisterGate will not verify that the given index is correct.
+func WithUnverifiedSolvableVar(solvableVar int) RegisterGateOption {
+	return func(settings *registerGateSettings) {
+		settings.noSolvableVarVerification = true
+		settings.solvableVar = solvableVar
+	}
+}
+
+// WithNoSolvableVar sets the gate as having no variable whose value can be uniquely determined from that of the other variables along with the gate's output.
+// RegisterGate will not check the correctness of this claim.
+func WithNoSolvableVar() RegisterGateOption {
+	return func(settings *registerGateSettings) {
+		settings.solvableVar = -1
+		settings.noSolvableVarVerification = true
+	}
+}
+
+// WithUnverifiedDegree sets the degree of the gate. RegisterGate will not verify that the given degree is correct.
+func WithUnverifiedDegree(degree int) RegisterGateOption {
+	return func(settings *registerGateSettings) {
+		settings.noDegreeVerification = true
+		settings.degree = degree
+	}
+}
+
+// WithDegree sets the degree of the gate. RegisterGate will return an error if the degree is not correct.
+func WithDegree(degree int) RegisterGateOption {
+	return func(settings *registerGateSettings) {
+		settings.degree = degree
+	}
+}
+
+// isAdditive returns whether x_i occurs only in a monomial of total degree 1 in f
+func (f GateFunction) isAdditive(i, nbIn int) bool {
+	// fix all variables except the i-th one at random points
+	// pick random value x1 for the i-th variable
+	// check if f(-, 0, -) + f(-, 2*x1, -) = 2*f(-, x1, -)
+	x := make(fr.Vector, nbIn)
+	x.MustSetRandom()
+	x0 := x[i]
+	x[i].SetZero()
+	in := slices.Clone(x)
+	y0 := f(in...)
+
+	x[i] = x0
+	copy(in, x)
+	y1 := f(in...)
+
+	x[i].Double(&x[i])
+	copy(in, x)
+	y2 := f(in...)
+
+	y2.Sub(&y2, &y1)
+	y1.Sub(&y1, &y0)
+
+	if !y2.Equal(&y1) {
+		return false // not linear
+	}
+
+	// check if the coefficient of x_i is nonzero and independent of the other variables (so that we know it is ALWAYS nonzero)
+	if y1.IsZero() { // f(-, x1, -) = f(-, 0, -), so the coefficient of x_i is 0
+		return false
+	}
+
+	// compute the slope with another assignment for the other variables
+	x.MustSetRandom()
+	x[i].SetZero()
+	copy(in, x)
+	y0 = f(in...)
+
+	x[i] = x0
+	copy(in, x)
+	y1 = f(in...)
+
+	y1.Sub(&y1, &y0)
+
+	return y1.Equal(&y2)
+}
+
+// fitPoly tries to fit a polynomial of degree less than degreeBound to f.
+// degreeBound must be a power of 2.
+// It returns the polynomial if successful, nil otherwise
```

Comment:
What would happen at the edge case where the gate is constant 0

Reply:
It panics. It's reasonable to fail in that case but I'll think about how to do it more gracefully.

Reply:
Made an explicit error for that case